### ISSUE TITLE
When picking up on files stored in Download for Android, FileData returns null

### ISSUE THREAD

**Issue by xinwanc** [jfversluis/FilePicker-Plugin-for-Xamarin-and-Windows#86]
## Expected Behavior
When I am selecting the file which is stored inside "Download", it should able to retrieves its file name and path
## Actual Behavior
When I am selecting the file which is stored inside "Download", it returns null.

## Steps to Reproduce the Problem

  1. When PickFile method is called, it displays the folders in Android
  2. Go to downloads, select a file
  3. It returns null
  4. However, it is working when selecting file from other folders in Android
## Specifications

  - NuGet package version: Beta 2.0.81
  - Platform(s):

**Comment by roberleitner** [comment]
I don't know if this is related but I figure this comment can't hurt. I've run into a similar issue that only seems to affect Android 8 or later. Files uploaded to the downloads folder through ADB via Android Studio are accessible but those downloaded through Chrome are not.

I cloned this project and see an IllegalArgumentException Unknown URI for the path specified which is something like "content://com.android.providers.downloads.documents/document/187".

Per a similar fix suggested with [this issue](https://github.com/flutter/flutter/issues/21863), changing "content://downloads/public_downloads" to "content://downloads/my_downloads" in `IOUtil.cs` fixes it for me and seems to work regardless of how the file was saved to the downloads folder.

**Comment by xinwanc** [comment]
Hi @ryanoberleitner thanks a lot for your reply! Therefore, it means this issue is occurred if the pdf file is downloaded from Chrome? I agree on what you said is because I selected a pdf file that is downloaded from Chrome into Downloads folder which end up returning null. Actually, this issue is not only occurred when selecting the file that is downloaded from Google Chrome but also on the file that is downloaded from Google Drive for me. Anyway, thanks for your assistance and may I know are you deciding to create pull request for this fix?

**Comment by roberleitner** [comment]
@xinwanc - I'm not sure the implications of changing public_downloads to my_downloads. There are comments on the issue I linked saying it doesn't always resolve the issue. I'm having a hard time finding any official documentation on these folders. I wouldn't suggest creating a pull request until there is more information available.

**Comment by kdongare** [comment]
I am also facing the same issue for pdf upload as it returns null if the file is downloaded from chrome.

**Comment by suryoaji** [comment]
> 
> 
> I am also facing the same issue for pdf upload as it returns null if the file is downloaded from chrome.

so do i, this is happened only on Android Oreo or above

does it need more permission to be added in androidmanifest or etc?

**Comment by xinwanc** [comment]
@kdongare  @suryoaji it also returns null if the file is downloaded from Google Drive as well for me. May I know do both of you facing the same issue while choosing the file downloaded from Google Drive?

**Comment by kdongare** [comment]
@xinwanc yes it does return null if the file is downloaded from Google drive.

**Comment by xinwanc** [comment]
@kdongare thanks for your prompt reply. According to @ryanoberleitner, his resolution is changing public_downloads to my_downloads but it doesn't completely resolve this issue as there are comments on the issue that he linked saying it doesn't always resolve the issue. @jfversluis can you advise on this matter? I am looking forward to receive your reply as soon as possible. Thank you.

**Comment by suryoaji** [comment]
For me, the best approach for getting path, i will get physical file from URI as InputStream, `ContentResolver.openInputStream()` allow you to access the contents of a file without knowing its real path.
Then write InputStream as a File (contains cache directory inside project package)

**Comment by suryoaji** [comment]
> 
> 
> @kdongare @suryoaji it also returns null if the file is downloaded from Google Drive as well for me. May I know do both of you facing the same issue while choosing the file downloaded from Google Drive?

yes, need to be downloaded manually (code implemented)

**Comment by xinwanc** [comment]
> For me, the best approach for getting path, i will get physical file from URI as InputStream, `ContentResolver.openInputStream()` allow you to access the contents of a file without knowing its real path.
> Then write InputStream as a File (contains cache directory inside project package)

@suryoaji may I know does this solve the issue?

**Comment by suryoaji** [comment]
Nope, i just try another way to get the file path with relocated it in the cache storage. So getting path issue should be solved.

**Comment by xinwanc** [comment]
I am sorry @suryoaji. May I know do you mean that when retrieving the file from google chrome or google drive you try to store it into the cache storage then retrieve the file path from the cache storage?

**Comment by vividos** [comment]
Sorry, my PR commit text closed this issue. But let me explain how picking on Android works. Every app has it's own process and internal storage space, and no app can access other app's data. There also is a public storage space, e.g. the file system. And there are some "content providers" that can also provide you with "files". The Android download manager is one such service, as is the OneDrive app and the Google Drive app.

The downloaded/cloud stored files are stored somewhere and the picker returns a "content URI" of the form "content://xxx/yyy". The Android class ContentResolver can be queried for file name, path name and a file stream. Not every content provider returns every info. When the file is stored in a database (or is retrieved from the interne) there may not even be an actual path name.

The best solution on Android therefore is to rely on that the FileData property DataArray is filled with the file contents of that file you have selected.

In a next version of the plugin we will integrate the access to the underlying Stream object that is read from, in order to be able to read large files or parse the file on the fly, without storing it somewhere.

tl;dr: On Android it's not always that easy to get an actual path name to a picked file; use the DataArray for now, or wait for accessing the Stream object.

**Comment by roberleitner** [comment]
thank you @vividos. any chance we can get a nuget beta release with this and the other changes you've made recently?

**Comment by vividos** [comment]
I'm merging some iOS fixes and then I'm figuring out how to create the NuGet package. Meanwhile you can get the CI build of the NuGet package from AppVeyor: https://ci.appveyor.com/project/jfversluis/filepicker-plugin-for-xamarin-and-windows-5pvwc/build/artifacts

**Comment by jfversluis** [comment]
NuGet package is now available. Please let us know if it works now.

**Comment by kdongare** [comment]
@jfversluis, Just now verified it on Pixel, Samsung S8 and Moto E device and it is working as expected.
Thanks for the quick update. 👍

**Comment by vividos** [comment]
You're welcome!

**Comment by xinwanc** [comment]
@vividos and @jfversluis  thank you very much

**Comment by vividos** [comment]
You're welcome!

**Comment by xinwanc** [comment]
Hi @vividos regarding selecting files from "Downloads" in Android. May I know that why it returns this file path:
![result](https://user-images.githubusercontent.com/38121289/47776945-89818100-dd2e-11e8-8f2c-ac6a979f7727.PNG)
As I want to invoke Device.OpenUri(new Uri(url)) to open the file, it can open the pdf viewer but I am unable to view content of the file. As for "url" I parsed in the file path.

**Comment by vividos** [comment]
Android uses content Uris to pass data between apps, and ContentResolver can be used to get the data from the content provider. About the external app... It seems the external app can't handle the link. For PDFs I saw this behavior with the OneDrive PDF viewer. Did you try out the Google Drive PDF viewer also?

**Comment by xinwanc** [comment]
I think i try out with Google Drive PDF viewer as default. Because when i invoke Device.OpenUri(new Uri(url)) it opens the file using Google Drive PDF viewer. I think because my file is downloaded from Google Chrome so it set Google Drive PDF viewer as default? I have find out that I manage to open the files that are downloaded from cloud storage, but I am unable to open the files that are downloaded from web browsers(I have checked both Google Chrome and Samsung Internet browsers). May I know is there any method to resolve it?

**Comment by vividos** [comment]
I'll have to try this out with the sample app. I'll get back to you

**Comment by xinwanc** [comment]
@vividos  sure thank you :)

**Comment by xinwanc** [comment]
Hi @vividos . I hope you are doing well. I have used the latest release of the plugin to test the file path when selecting files from "Downloads" in Android but it still throw the same issue as discussed on 31 October. 

> Hi @vividos regarding selecting files from "Downloads" in Android. May I know that why it returns this file path:
> ![result](https://user-images.githubusercontent.com/38121289/47776945-89818100-dd2e-11e8-8f2c-ac6a979f7727.PNG)
> As I want to invoke Device.OpenUri(new Uri(url)) to open the file, it can open the pdf viewer but I am unable to view content of the file. As for "url" I parsed in the file path.

**Comment by vividos** [comment]
I think the URL may only be valid for your app and also only after picking the file. If you want to open the file, be sure to download it, store it in your local storage and implement a ContentProvider to share it with another app. I think I saw the Xamarin.Essentials project on working on such a thing (GH-130).

**Comment by ralpholazo24** [comment]
> Hi @vividos . I hope you are doing well. I have used the latest release of the plugin to test the file path when selecting files from "Downloads" in Android but it still throw the same issue as discussed on 31 October.
> 
> > Hi @vividos regarding selecting files from "Downloads" in Android. May I know that why it returns this file path:
> > ![result](https://user-images.githubusercontent.com/38121289/47776945-89818100-dd2e-11e8-8f2c-ac6a979f7727.PNG)
> > As I want to invoke Device.OpenUri(new Uri(url)) to open the file, it can open the pdf viewer but I am unable to view content of the file. As for "url" I parsed in the file path.

Hi @xinwanc, i try to pick a file from my download folder and it still returns null. But if I try on the SD Card file path, I get the same value of yours. Can you help me to figure this out why?

**Comment by vividos** [comment]
From my experience I get the impression that the download manager downloads files in the download folder and keeps additional data about the file in some database, e.g. the MIME type. Now when you move away the file manually, e.g. with Total Commander, it still thinks the file is there and let's you pick the file, even when the actual file is already gone. Maybe that's the case here.
In my app I just show the user a message that the file couldn't be picked.
